apps/aecore/src/aec_chain_state.erl

```diff
@@ -0,0 +1,458 @@
+
+determine_chain_relation(Node, State) ->
+    Height   = node_height(Node),
+    Hash     = Node#node.hash,
+    PrevHash = prev_hash(Node),
+    assert_previous_height(Height, PrevHash, State),
+    case get_top_header_hash(State) of
+        PrevHash -> new_top;
+        undefined when Height > 0 ->
+            off_chain; %% No proper chain yet.
+        undefined when Height =:= 0 ->
+            new_top; %% A genesis block. TODO: Should this be checked?
+        Hash ->
+            in_chain; %% This is the top header
+        TopHash when is_binary(TopHash), Height =:= 0 ->
+            %% A new genesis block. TODO: This
+            error(new_genesis_block_nyi);
+        TopHash when is_binary(TopHash) ->
+            case is_node_in_main_chain(Node, State) of
+                true -> in_chain;
+                false ->
+                    TopNode = blocks_db_get(TopHash, State),
+                    case find_fork_point(TopNode, Node, State) of
+                        not_found -> off_chain;
+                        ForkNode -> {fork, ForkNode}
+                    end
+            end
+    end.
+
+assert_previous_height(Height, PrevHash, State) ->
+    case blocks_db_find(PrevHash, State) of
+        {ok, Node} ->
+            case node_height(Node) =:= (Height - 1) of
+                true -> ok;
+                false -> internal_error(height_inconsistent_with_previous_hash)
+            end;
+        error -> ok
+    end.
+
+find_fork_point(Node1, Node2, State) ->
+    Height1 = node_height(Node1),
+    Height2 = node_height(Node2),
+    find_fork_point(Node1, Height1, Node2, Height2, State).
+
+find_fork_point(Node, Height, Node, Height,_State) -> Node;
+find_fork_point(_Node1, Height,_Node2, Height,_State) when Height =:= 0 ->
+    not_found;
+find_fork_point(Node1, Height1, Node2, Height2, State) when Height1 > Height2,
+                                                            Height1 > 0 ->
+    PrevHash = prev_hash(Node1),
+    case blocks_db_find(PrevHash, State) of
+        error -> not_found;
+        {ok, NewNode} ->
+            NewHeight = node_height(NewNode),
+            find_fork_point(NewNode, NewHeight, Node2, Height2, State)
+    end;
+find_fork_point(Node1, Height1, Node2, Height2, State) when Height2 >= Height2,
+                                                            Height2 > 0 ->
+    PrevHash = prev_hash(Node2),
+    case blocks_db_find(PrevHash, State) of
+        error -> not_found;
+        {ok, NewNode} ->
+            NewHeight = node_height(NewNode),
+            find_fork_point(Node1, Height1, NewNode, NewHeight, State)
+    end.
+
+find_new_header_top_from_node(Node, State) ->
+    Height = node_height(Node),
+    Hash = hash(Node),
+    case children(blocks_db_find_at_height(Height + 1, State), Hash) of
+        [] -> [Node];
+        [NextNode] -> find_new_header_top_from_node(NextNode, State);
+        [_|_]  = List ->
+            lists:flatten([find_new_header_top_from_node(N, State)
```

Comment:
why not `lists:flatmap(fun(N) -> find_new_header_top_from_node(N, State) end, List)`?

Reply:
Sure.